messaginginapp/src/androidTest/java/io/customer/messaginginapp/support/GistEngineWebViewClientInterceptor.kt

```diff
@@ -0,0 +1,140 @@
+import io.customer.messaginginapp.support.GistEngineMessageDriver.Callback
+import io.customer.messaginginapp.support.GistEngineMessageDriver.MessageState
+import java.util.concurrent.ConcurrentHashMap
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.TestScope
+
+/**
+ * This class is responsible for intercepting WebViewClient events and simulate
+ * behavior of actual in-app messaging loaded in WebView by leveraging GistEngineMessageDriver.
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+class GistEngineWebViewClientInterceptor(
+    private val testCoroutineScope: TestScope
+) : EngineWebViewClientInterceptor, Callback, GistListener {
+    private val gson: Gson = Gson()
+    private val messageDriversMap: ConcurrentHashMap<String, GistEngineMessageDriver> = ConcurrentHashMap()
+
+    fun getMessageDriver(message: Message): GistEngineMessageDriver {
+        return messageDriversMap.getOrPut(message.messageId) { GistEngineMessageDriver(message) }
+    }
+
+    // Used to reset the message driver for cases where the message is dismissed or cancelled
+    private fun clearMessageDriver(message: Message) {
+        messageDriversMap.remove(message.messageId)
+    }
+
+    fun reset() {
+        messageDriversMap.clear()
+    }
+
+    override fun onPageFinished(view: WebView, url: String?) {
+        val uri = Uri.parse(url) ?: return
+        val options = uri.getQueryParameter("options") ?: return
+
+        val message = decodeOptionsString(options)
+        val messageDriver = getMessageDriver(message)
+
+        messageDriver.setup(callback = this, webView = view)
+        messageDriver.onStateComplete(MessageState.HTML_LOADED)
+    }
+
+    override fun handleBootstrapped(driver: GistEngineMessageDriver) {
+        val engineMessage = EngineWebMessage(
+            gist = EngineWebEvent(
+                instanceId = driver.message.instanceId,
+                method = "bootstrapped",
+                parameters = emptyMap()
+            )
+        )
+
+        postMessage(
+            driver = driver,
+            engineMessage = engineMessage,
+            messageState = MessageState.BOOTSTRAPPED
+        )
+    }
+
+    override fun handleRouteLoaded(driver: GistEngineMessageDriver) {
+        val route = driver.message
+            .messageId
+            .takeIf { id -> id.isNotBlank() }
+        if (route == null) {
+            postMessage(
+                driver = driver,
+                engineMessage = null,
+                messageState = MessageState.COMPLETED
+            )
+            return
+        }
+
+        val engineMessage = EngineWebMessage(
+            gist = EngineWebEvent(
+                instanceId = driver.message.instanceId,
+                method = "routeLoaded",
+                parameters = buildMap {
+                    put("route", route)
+                }
```

Comment:
```suggestion
                parameters = mapOf("route" to route)
```

`buildMap` seems overkill considering the use-case? 😄